src/client/dfs/dfs.c

```diff
@@ -7446,3 +7442,405 @@ dfs_obj_fix_type(dfs_t *dfs, dfs_obj_t *parent, const char *name)
+			D_GOTO(err, rc = daos_der2errno(rc));
+		rc = daos_filter_add(&dpipe->pipef, &dpipe->akey1_ft);
+		if (rc)
+			D_GOTO(err, rc = daos_der2errno(rc));
+		rc = daos_filter_add(&dpipe->pipef, &dpipe->const1_ft);
+		if (rc)
+			D_GOTO(err, rc = daos_der2errno(rc));
+		rc = daos_filter_add(&dpipe->pipef, &dpipe->const2_ft);
+		if (rc)
+			D_GOTO(err, rc = daos_der2errno(rc));
+	}
+
+	if (flags & DFS_FILTER_NEWER && flags & DFS_FILTER_NAME) {
+		rc = daos_filter_add(&dpipe->pipef, &dpipe->and_ft);
+		if (rc)
+			D_GOTO(err, rc = daos_der2errno(rc));
+	}
+
+	if (flags & DFS_FILTER_NAME) {
+		rc = daos_filter_add(&dpipe->pipef, &dpipe->like_ft);
+		if (rc)
+			D_GOTO(err, rc = daos_der2errno(rc));
+		rc = daos_filter_add(&dpipe->pipef, &dpipe->dkey_ft);
+		if (rc)
+			D_GOTO(err, rc = daos_der2errno(rc));
+		rc = daos_filter_add(&dpipe->pipef, &dpipe->const0_ft);
+		if (rc)
+			D_GOTO(err, rc = daos_der2errno(rc));
+	}
+
+	if (flags & DFS_FILTER_NEWER) {
+		rc = daos_filter_add(&dpipe->pipef, &dpipe->gt_ft);
+		if (rc)
+			D_GOTO(err, rc = daos_der2errno(rc));
+		rc = daos_filter_add(&dpipe->pipef, &dpipe->akey2_ft);
+		if (rc)
+			D_GOTO(err, rc = daos_der2errno(rc));
+		rc = daos_filter_add(&dpipe->pipef, &dpipe->const3_ft);
+		if (rc)
+			D_GOTO(err, rc = daos_der2errno(rc));
+	}
+
+	rc = daos_pipeline_add(&dpipe->pipeline, &dpipe->pipef);
+	if (rc)
+		D_GOTO(err, rc = daos_der2errno(rc));
+
+	*_dpipe = dpipe;
+	return 0;
+err:
+	printf("failed to create pipeline. rc = %d\n", rc);
+	D_FREE(dpipe);
+	return rc;
+}
+
+int
+dfs_pipeline_destroy(dfs_pipeline_t *dpipe)
+{
+	if (dpipe->pipeline.num_filters)
+		D_FREE(dpipe->pipeline.filters);
+	D_FREE(dpipe);
+	return 0;
+}
+
+int
+dfs_readdir_with_filter(dfs_t *dfs, dfs_obj_t *obj, dfs_pipeline_t *dpipe, daos_anchor_t *anchor,
+			uint32_t *nr, struct dirent *dirs, daos_obj_id_t *oids, daos_size_t *csize,
+			uint64_t *nr_scanned)
+{
+	daos_iod_t	iod;
+	daos_key_desc_t	*kds;
+	d_sg_list_t	sgl_keys, sgl_recs;
+	d_iov_t		iov_keys, iov_recs;
+	char		*buf_keys = NULL, *buf_recs = NULL;
+	daos_recx_t	recxs[4];
+	uint32_t	nr_iods, nr_kds, key_nr, i;
+	daos_size_t	record_len;
+	int		rc = 0;
+
+	if (dfs == NULL || !dfs->mounted)
+		return EINVAL;
+	if (obj == NULL || !S_ISDIR(obj->mode))
+		return ENOTDIR;
+	if (*nr == 0)
+		return 0;
+	if (dpipe == NULL || dirs == NULL || anchor == NULL)
+		return EINVAL;
+
+	/* IOD to retrieve the mode_t and the ctime */
+	iod.iod_nr	= 2;
+	iod.iod_size	= 1;
+	recxs[0].rx_idx	= MODE_IDX;
+	recxs[0].rx_nr	= sizeof(mode_t);
+	recxs[1].rx_idx	= CTIME_IDX;
+	recxs[1].rx_nr	= sizeof(time_t);
+	iod.iod_recxs	= recxs;
+	iod.iod_type	= DAOS_IOD_ARRAY;
+	d_iov_set(&iod.iod_name, INODE_AKEY_NAME, sizeof(INODE_AKEY_NAME) - 1);
+	record_len = recxs[0].rx_nr + recxs[1].rx_nr;
+
+	if (oids) {
+		iod.iod_nr ++;
+		recxs[2].rx_idx	= OID_IDX;
+		recxs[2].rx_nr	= sizeof(daos_obj_id_t);
+		record_len += recxs[2].rx_nr;
+	}
+	if (csize) {
+		iod.iod_nr ++;
+		recxs[3].rx_idx	= CSIZE_IDX;
```

Comment:
question: could "oids" be NULL? If "oids" is NULL, shall we still use "recxs[3]" or "recxs[iod.iod_nr - 1]" in these three lines?

Reply:
yes indeed it can be null since it's optional. ill fix that, thx.